src/realm/object-store/c_api/app.cpp

```diff
@@ -385,6 +385,11 @@ RLM_API realm_user_t* realm_app_get_current_user(const realm_app_t* app) noexcep
 RLM_API bool realm_app_get_all_users(const realm_app_t* app, realm_user_t** out_users, size_t capacity, size_t* out_n)
 {
     return wrap_err([&] {
+        const size_t users_size = (*app)->all_users().size();
+        set_array_size(out_n, users_size);
```

Comment:
This is prone to race conditions because we first get `all_users().size()`, then we get `all_users()`, meaning the collection may have changed between the two invocations. I think we should only ever get the collection once and use the copied vector we received. Something like:

```cpp
auto users = (*app)->all_users();
set_array_size(out_n, users.size());
if (capacity < users.size() || users.size() == 0) {
    return;
}

// If SDKs provided a non-zero capacity, they must have provided a non-null out_users pointer
assert(out_users);

OutBuffer<realm_user_t> buf(out_users);
for (const auto& user : users) {
    buf.emplace(user);
}
buf.release(out_n);
```

Reply:
I don't think I understand your proposal. 
We are getting a copy of user's `vector`, calling `all_users()`. If the number of users changes in between the `all_users()` and `set_array_size()`, you might still have a race condition. I don't think your code solves the issue. 
It is still prone to races.

Reply:
When we get `all_users()`, we get a copy of the vector inside a mutex. Once the method returns, we should not see size changes on the vector.

Reply:
Understood @nirinchev suggestion.